File: src/pl/swmud/ns/swaedit/gui/ResetFlagsWidget.java
package pl.swmud.ns.swaedit.gui;

import java.util.List;

import pl.swmud.ns.swaedit.core.FlagsWrapper;
import pl.swmud.ns.swaedit.core.IFlagsSetter;
import pl.swmud.ns.swaedit.core.ResetWrapper;
import pl.swmud.ns.swaedit.flags.Flag;

import com.trolltech.qt.gui.QMessageBox;
import com.trolltech.qt.gui.QWidget;

class ResetFlagsWidget extends QWidget {
    
    private Ui_ValueFlagsWidget ui = new Ui_ValueFlagsWidget();
    private boolean canEdit = false;
    private List<Flag> flagsTree;
    private Long flags;
    private ResetWrapper wrapper;
    private String title;
    public Signal1<String> textChanged = new Signal1<String>();
    public Signal0 clicked = new Signal0();
    
    public ResetFlagsWidget(ResetWrapper wrapper) {
        super(SWAEdit.ref);
        ui.setupUi(this);
        this.wrapper = wrapper;
        canEdit = true;
    }
    
    @SuppressWarnings("unused")
    private void on_edit_textChanged(String str) {
        if (!canEdit) {
            return;
        }

        try {
            setFWFlags(Long.parseLong(str));
            wrapper.setCurrentValue(flags);
            SWAEdit.ref.updateResetNavigatorText(wrapper);
            SWAEdit.ref.setModified();
        } catch (NumberFormatException e) {
            if( !str.isEmpty() ) {
                QMessageBox.critical(null, "Invalid Number Value", title+" must have number value!");
                setFWFlags(wrapper.getCurrentValue());
            }
        }
    }

    @SuppressWarnings("unused")
    private void on_button_clicked() {
        if (!canEdit) {
            return;
        }
        clicked.emit();
    }
    
    Long getFlags() {
        return flags;
    }

    void setFWFlags(Long flags) {
        canEdit = false;
        this.flags = flags;
        ui.edit.setText(flags.toString());
        canEdit = true;
    }

    List<Flag> getFlagsTree() {
        return flagsTree;
    }

    void setFlagsTree(List<Flag> flagsTree) {
        this.flagsTree = flagsTree;
    }

    public void setTitle(String title) {
        this.title = title;
    }
    
    void showFlagsWidget() {
        new FlagsWidget(flagsTree,new FlagsWrapper(flags,
                new IFlagsSetter() {
                    public void setFlags(Long flagsValue) {
                        setFWFlags(flagsValue);
                        wrapper.setCurrentValue(flagsValue);
                        SWAEdit.ref.updateResetNavigatorText(wrapper);
                        SWAEdit.ref.setModified();
                    }
        }),title).show();
    }

}

File: src/gui/ResetFlagsWidget.cpp
#include "gui/ResetFlagsWidget.h"
#include "gui/FlagsWidget.h"
#include "gui/MainWindow.h"
#include "ui_ValueFlagsWidget.h"

ResetFlagsWidget::ResetFlagsWidget(int argIndex,
                                   qint64 flags,
                                   const QList<FlagDef> &flagsTree,
                                   AreaReset *reset,
                                   const QString &title,
                                   MainWindow *mainWindow,
                                   QWidget *parent)
    : QWidget(parent)
    , ui(new Ui::ValueFlagsWidget)
    , argIndex_(argIndex)
    , flags_(flags)
    , flagsTree_(flagsTree)
    , reset_(reset)
    , title_(title)
    , mainWindow_(mainWindow)
{
    ui->setupUi(this);

    ui->edit->setText(QString::number(flags_));

    connect(ui->edit, &QLineEdit::textChanged, this, &ResetFlagsWidget::onEditTextChanged);
    connect(ui->button, &QPushButton::clicked, this, &ResetFlagsWidget::onButtonClicked);
}

ResetFlagsWidget::~ResetFlagsWidget()
{
    delete ui;
}

void ResetFlagsWidget::setFWFlags(qint64 flags)
{
    canEdit_ = false;
    flags_ = flags;
    ui->edit->setText(QString::number(flags_));
    canEdit_ = true;
}

void ResetFlagsWidget::onEditTextChanged(const QString &text)
{
    if (!canEdit_) return;
    emit textChanged(text);
}

void ResetFlagsWidget::onButtonClicked()
{
    emit clicked();
    showFlagsWidget();
}

void ResetFlagsWidget::showFlagsWidget()
{
    bool ok;
    qint64 currentValue = ui->edit->text().toLongLong(&ok);
    if (!ok) currentValue = flags_;

    FlagsWidget *fw = new FlagsWidget(flagsTree_, currentValue, title_, false, this);
    connect(fw, &FlagsWidget::flagsAccepted, this, [this](qint64 value) {
        flags_ = value;

        // Update the reset arg value
        if (reset_) {
            switch (argIndex_) {
            case 0: reset_->extra = static_cast<int>(value); break;
            case 1: reset_->arg1 = static_cast<int>(value); break;
            case 2: reset_->arg2 = static_cast<int>(value); break;
            case 3: reset_->arg3 = static_cast<int>(value); break;
            case 4: reset_->arg4 = static_cast<int>(value); break;
            }
        }

        setFWFlags(value);

        if (mainWindow_) {
            mainWindow_->setModified();
            mainWindow_->updateResetNavigatorText();
        }
    });

    fw->show();
}
